File: a_code/A015_企查查_Hmac/qichacha.js
/*
js位置
var u = function(e) {
    var t = e.url.replace(e.baseURL, "")
      , n = o.default.stringify(e.params || {});
    n && (t += (-1 === t.indexOf("?") ? "?" : o.default.options.delimiter || "&") + n),
    t = t.toLowerCase();
    var i = (0,
    a.default)(t, e.data)
      , l = (0,
    r.default)(t, e.data, (0,
    s.default)());
    e.headers[i] = l
};
* */

// t = t.toLowerCase();
// var i = a.default(t, e.data),
//     l = r.default(t, e.data, s.default());

var CryptoJS = require('crypto-js');

// ===================== 分析 t ========================================
// api + 参数 拼接， 转小写而来
t = '/api/datalist/touzilist?keyno=5dffb644394922f9073544a08f38be9f&pageindex=2';


// ================== 分析 i ============================================
// var i = a.default(t, e.data)

// r函数中 o.default 是写死的
r_o = {};  // 这里指代r函数中的o
r_o.default = {
    "n": 20,
    "codes": {
        "0": "W",
        "1": "l",
        "2": "k",
        "3": "B",
        "4": "Q",
        "5": "g",
        "6": "f",
        "7": "i",
        "8": "i",
        "9": "r",
        "10": "v",
        "11": "6",
        "12": "A",
        "13": "K",
        "14": "N",
        "15": "k",
        "16": "4",
        "17": "L",
        "18": "1",
        "19": "8"
    }
};

// s中: a.default(t)
var r = function() {
    for (var e = (arguments.length > 0 && void 0 !== arguments[0] ? arguments[0] : "/").toLowerCase(), t = e + e, n = "", i = 0; i < t.length; ++i) {
        var a = t[i].charCodeAt() % r_o.default.n;
        n += r_o.default.codes[a]
    }
    return n
};

function get_i(t) {
    // i = a.default(t, e.data) // e.data为 undefined
    var s = function() {
        var e = arguments.length > 1 && void 0 !== arguments[1] ? arguments[1] : {}
          , t = (arguments.length > 0 && void 0 !== arguments[0] ? arguments[0] : "/").toLowerCase()
          , n = JSON.stringify(e).toLowerCase();
        // return o.default(t + n, a.default(t)).toLowerCase().substr(8, 20)
        // o.default， 是一个标准 HmacSHA512 加密,可用算法包模拟
        return CryptoJS.HmacSHA512(t + n, r(t)).toString().toLowerCase().substr(8, 20)
    };

    // headers的key
    i = s(t, undefined);
    return i;
}

console.log('i:', get_i(t));


// ======================= 分析 l =============================
// l = r.default(t, e.data, s.default());

// s.default(), 就是window.tid， 是一个定值；
// var _default = function _default() {
//     var list = ["w", "i", "n", "d", "o", "w", ".", "t", "i", "d"];
//     return eval(list.join(""))
// };

function get_l(t){
    // s.default()
    tid = 'd8e943c06de83114b7927251b675b4a1';

    // r.default
    var s2 = function() {
        var e = arguments.length > 1 && void 0 !== arguments[1] ? arguments[1] : {}
          , t = arguments.length > 2 && void 0 !== arguments[2] ? arguments[2] : ""
          , n = (arguments.length > 0 && void 0 !== arguments[0] ? arguments[0] : "/").toLowerCase()
          , i = JSON.stringify(e).toLowerCase();
        // return o.default(n + "pathString" + i + t, a.default(n))
        return CryptoJS.HmacSHA512(n + "pathString" + i + t, r(n)).toString()
    };

    l = s2(t, undefined, tid);
    return l;
}

console.log('l:', get_l(t));


// 封装入口
function get_k_v(t) {
    key = get_i(t);
    value = get_l(t);
    return [key, value]
}

// console.log(get_k_v(t));

File: a_code/A015_企查查_Hmac/qichacha.py
import requests

from utils.ut import compile_js, format_query_params


"""分析：
请求头有一个 加密参数，key 和 value 都是加密参数,且：
- 每页的 k，v 是固定的，根据api和参数加密而来；

加密方式使用: HmacSHA512
"""

def get_data():
    url = 'https://www.qcc.com/api/datalist/touzilist'
    params = {
        'keyNo': '5dffb644394922f9073544a08f38be9f',  # 应该是公司id， 可从列表api取值
        'pageIndex': 5  # 页码
    }
    t = url.replace('https://www.qcc.com', '') + '?' + format_query_params(params).lower()
    exe = compile_js('qichacha.js')
    k, v = exe.call('get_k_v', t)

    headers = {
        # "accept": "application/json, text/plain, */*",
        # "accept-encoding": "gzip, deflate, br",
        # "accept-language": "zh-CN,zh;q=0.9,en;q=0.8",
        # "b26315bd778a64e24856": "1eeb21d5bebd3d098ecf6c2a58f3ba6e2aa4dc7bc479362ea9c63466ffd2fea0309f244ca655518aade29020d1b29ee9fbe893ad586087d59a3b42ab4d5f4584",
        k: v,
        # "cache-control": "no-cache",
        # TODO: 暂未分析是否会改变
        "cookie": "_uab_collina=162463350927257430988502; qcc_did=b79eac63-19b3-4d62-908e-4b18b3bb992c; UM_distinctid=1867c3f28ff4dd-017161df0bc1d9-26031951-384000-1867c3f2900ceb; MQCCSESSID=161f9b6a144fab76c04ea01480; QCCSESSID=841ab184084288c4d863b60287; CNZZDATA1254842228=1522912875-1677120519-https%253A%252F%252Fwww.baidu.com%252F%7C1677746971; acw_tc=3b2550a216777473291188009ea02abc239639e2c0cc183b6e0f8ab97c",
        # "pragma": "no-cache",
        # "referer": "https://www.qcc.com/firm/5dffb644394922f9073544a08f38be9f.html",
        # "sec-ch-ua": "\"Chromium\";v=\"110\", \"Not A(Brand\";v=\"24\", \"Google Chrome\";v=\"110\"",
        # "sec-ch-ua-mobile": "?1",
        # "sec-ch-ua-platform": "\"Android\"",
        # "sec-fetch-dest": "empty",
        # "sec-fetch-mode": "cors",
        # "sec-fetch-site": "same-origin",
        "user-agent": "Mozilla/5.0 (Linux; Android 6.0; Nexus 5 Build/MRA58N) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/110.0.0.0 Mobile Safari/537.36",
        # "x-pid": "0f028790a20964f4f9d6ab5202253945",
        # "x-requested-with": "XMLHttpRequest"
    }

    resp = requests.get(url, params=params, headers=headers)
    res = resp.json()
    return res


if __name__ == '__main__':
    print(get_data())